pkg/cli/version/version.go

```diff
@@ -1,25 +1,184 @@
+	return &VersionOptions{
+		IOStreams: ioStreams,
+	}
+}
+
+// NewCmdVersion is copied from upstream NewCmdVersion with addition of OpenShift Server version info.
+// OpenShift Server version is output only if logged in to a cluster as an admin user.
+func NewCmdVersion(fullName string, f cmdutil.Factory, ioStreams genericclioptions.IOStreams) *cobra.Command {
+	o := NewVersionOptions(ioStreams)
 	cmd := &cobra.Command{
-		Use:   "version",
-		Short: "Display version",
-		Long:  "Display version",
+		Use:     "version",
+		Short:   i18n.T("Print the client and server version information"),
+		Long:    "Print the client and server version information for the current context",
+		Example: fmt.Sprintf(versionExample, fullName),
 		Run: func(cmd *cobra.Command, args []string) {
-			fmt.Fprintf(out, "%s %v\n", fullName, versionInfo)
+			cmdutil.CheckErr(o.Validate())
+			cmdutil.CheckErr(o.Complete(f, cmd))
+			cmdutil.CheckErr(o.Run())
 		},
 	}
-
+	cmd.Flags().BoolVar(&o.ClientOnly, "client", o.ClientOnly, "Client version only (no server required).")
+	cmd.Flags().BoolVar(&o.Short, "short", o.Short, "Print just the version number.")
+	cmd.Flags().StringVarP(&o.Output, "output", "o", o.Output, "One of 'yaml' or 'json'.")
 	return cmd
 }
+
+// Complete is copied from upstream version command with added clusteroperator client
+// to report OpenShift server version
+func (o *VersionOptions) Complete(f cmdutil.Factory, cmd *cobra.Command) error {
+	var err error
+	if o.ClientOnly {
+		return nil
+	}
+	o.discoveryClient, err = f.ToDiscoveryClient()
+	// if we had an empty rest.Config, continue and just print out client information.
+	// if we had an error other than being unable to build a rest.Config, fail.
+	if err != nil && !clientcmd.IsEmptyConfig(err) {
+		return err
+	}
+
+	clientConfig, err := f.ToRESTConfig()
+	if err != nil && !clientcmd.IsEmptyConfig(err) {
+		return err
+	}
+	if clientConfig != nil {
+		o.oClient, err = configv1client.NewForConfig(clientConfig)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Run is copied from upstream version command, with added OpenShift server version logic
+func (o *VersionOptions) Run() error {
+	var versionInfo Version
+	clientVersion := version.Get()
+	versionInfo.ClientVersion = &clientVersion
+
+	var serverErr error
+	var serverVersion *apimachineryversion.Info
+	if !o.ClientOnly {
+		if o.discoveryClient != nil {
+			// Always request fresh data from the server
+			o.discoveryClient.Invalidate()
+			serverVersion, serverErr = o.discoveryClient.ServerVersion()
+			versionInfo.ServerVersion = serverVersion
+		}
+		if o.oClient != nil {
+			var clusterOperator *configv1.ClusterOperator
+			clusterOperator, serverErr = o.oClient.ClusterOperators().Get("openshift-apiserver", metav1.GetOptions{})
+			// if error here, not logged in as admin, move on
+			if serverErr != nil {
+				switch {
+				case kerrors.IsForbidden(serverErr), kerrors.IsNotFound(serverErr):
+					serverErr = nil
```

Comment:
at least log the error at higher loglevel if you don't want to see it by default

Reply:
it will help us when we ask people for `oc --loglevel=X ...` to prove what went wrong for them

Reply:
ok, will do